### PR TITLE
H-21: Add `hash.dev/roadmap` page

### PR DESCRIPTION
**Pull request by benwerner01**
## 🌟 What is the purpose of this PR?

This PR adds the `/roadmap` page to the hash.dev site, which features a "technology tree" that displays a directed acyclic graph of upcoming and completed tasks and how they relate.

## 🔍 What does this change?

- upgrades TS packages to `5.1.6` so that the latest version of the types in the `d3-dag` package can be used

### 🚢 Has this modified a publishable library?

This PR:

 - [x] does not modify any publishable blocks or libraries, or modifications do not need publishing

### 📜 Does this require a change to the docs?

The changes in this PR:

 - [x] is internal and does not require a docs change

### 🕸️ Does this require a change to the Turbo Graph?

The changes in this PR:

 - [x] does not affect the execution graph

## ⚠️ Known issues

- Currently the zooming is not perfect, as it zooms slightly below where the mouse cursor is. I'm not entirely sure what's causing this, but it's likely to do with the calculations I've had to make to get zooming behaviour to both affect the SVG edges and the HTML nodes consistently.
- Because of how we are currently using Font Awesome icons in the hash.dev site, they are loaded client side and not being returned by the server as SVG elements part of the DOM. This means on slow internet connections, these can flash into view briefly moving elements on the page. This is being tracked as part of [this task](https://linear.app/hash/issue/H-171)

## 🐾 Next steps

- add animations between states in the graph (i.e. when different filters are used to filter the nodes)

## 🛡 What tests cover this?

None

## ❓ How to test this?

Open the deployment and go to the `/roadmap` page.

## 📹 Demo

https://github.com/hashintel/hash/assets/42802102/9979b9aa-a61c-42b7-824c-17bc56c5e3b6

### CODE REVIEW

File: apps/hashdotdev/src/pages/roadmap.page.tsx
Comment (nonparibus) [inline apps/hashdotdev/src/pages/roadmap.page.tsx:123]:
```suggestion
    name: "Internal Tools/Apps",
```

Updated here: https://www.figma.com/file/JSX4lGEb5N94q0XlvqzN0O?node-id=1858:15241&mode=design#495627083

**Comment by nonparibus** [comment]
_Left comments on the roadmap page of the Vercel deployment._

File: apps/hashdotdev/src/pages/roadmap.page.tsx
Comment (CiaranMn) [inline apps/hashdotdev/src/pages/roadmap.page.tsx:50]:
```suggestion
            We’ve mapped these features to use cases, so if you’re interested in
```

File: apps/hashdotdev/src/pages/roadmap.page.tsx
Comment (CiaranMn) [inline apps/hashdotdev/src/pages/roadmap.page.tsx:45]: copy suggestion for @nonparibus 
```suggestion
            Here you’ll find the features we’ve built and are building, information
```

File: apps/hashdotdev/src/pages/roadmap.page.tsx
Comment (CiaranMn) [inline apps/hashdotdev/src/pages/roadmap.page.tsx:66]: @nonparibus for view
```suggestion
            <strong>Quick links</strong>
```

File: apps/hashdotdev/src/components/anchor-tag.tsx
Comment (CiaranMn) [inline apps/hashdotdev/src/components/anchor-tag.tsx:6]: Personally I'd name this `AnchorFragmentTag` or something more specific (maybe even throw 'hidden' in there), since any `<a>` is an anchor tag

File: apps/hashdotdev/src/pages/roadmap.page.tsx
Comment (CiaranMn) [inline apps/hashdotdev/src/pages/roadmap.page.tsx:198]: Anything stopping us from doing this now?

File: apps/hashdotdev/src/pages/roadmap/technology-tree.tsx
Comment (CiaranMn) [inline apps/hashdotdev/src/pages/roadmap/technology-tree.tsx:326]: Maybe we can make this an `onClick` instead or consider some other means of selecting/deselecting nodes? It's hard to navigate while keeping the info tooltip hovered. cc @nonparibus

**Review by CiaranMn — COMMENTED** [review]
Thanks @benwerner01, great work! The tech tree is very cool and looks fantastic

I left a few comments on the code – other than that, we should do a pass on the data before merging it, to make sure everything is correct, and to assign the proper use cases to everything.

File: apps/hashdotdev/package.json
Comment (benwerner01) [reply apps/hashdotdev/package.json:52]: Agree it should remain in dev dependencies, but the reason I added it here is because in all other TS packages we are putting it into the dependencies, and making it consistent allowed me to use the existing `fix:dependency-version-consistency` script to programmatically update all other `package.json` files.

I could also move it into the dev dependencies for all packages though?

File: apps/hashdotdev/src/theme/components/utils/mui-css-baseline-theme-options.ts
Comment (benwerner01) [reply apps/hashdotdev/src/theme/components/utils/mui-css-baseline-theme-options.ts:203]: It didn't take effect (i.e. the scrolling wasn't smooth) with the `!important` flag. Not sure why, but we're also using `scroll-behavior: smooth !important;` in the BP site.

File: apps/hashdotdev/src/pages/roadmap.page.tsx
Comment (benwerner01) [reply apps/hashdotdev/src/pages/roadmap.page.tsx:50]: This copy is from the figma. @nonparibus would you like me to make this change? There's also another place we're using "use-case" in copy.

File: apps/hashdotdev/src/pages/roadmap.page.tsx
Comment (benwerner01) [reply apps/hashdotdev/src/pages/roadmap.page.tsx:66]: This copy is also from the Figma, @nonparibus would you like me to make this change?

File: apps/hashdotdev/src/components/anchor-tag.tsx
Comment (benwerner01) [reply apps/hashdotdev/src/components/anchor-tag.tsx]: Fair point, have renamed to `HiddenAnchorFragmentTag` in https://github.com/hashintel/hash/pull/2747/commits/23b5abb89eef42cf16cb6e1d11bcddbcebee9a39

File: apps/hashdotdev/src/pages/roadmap.page.tsx
Comment (benwerner01) [reply apps/hashdotdev/src/pages/roadmap.page.tsx]: Addressed in https://github.com/hashintel/hash/pull/2747/commits/20b401d4d3479cd015d5667d25e5c657b51595cc

image

File: apps/hashdotdev/src/pages/roadmap/technology-tree.tsx
Comment (benwerner01) [reply apps/hashdotdev/src/pages/roadmap/technology-tree.tsx:362]: I'm not sure, it didn't make sense to me either but it was somehow occurring when selecting different filters so this was how I addressed it.

And yes this removes duplicate links, which prevents there from being duplicate react keys because those are derived from the `source` and `target` IDs of a link.

File: apps/hashdotdev/src/pages/roadmap/technology-tree-data.tsx
Comment (benwerner01) [reply apps/hashdotdev/src/pages/roadmap/technology-tree-data.tsx:10]: I can remove this yes, the reason why I added it was to allow @nonparibus to use bold/italics in the copy. If we don't for-see needing this I can make this a `.ts` file and constrain the type.

File: apps/hashdotdev/src/pages/roadmap.page.tsx
Comment (nonparibus) [reply apps/hashdotdev/src/pages/roadmap.page.tsx]: Can we vertically center them too? E.g. website building looks off

File: apps/hashdotdev/src/pages/roadmap/technology-tree.tsx
Comment (benwerner01) [reply apps/hashdotdev/src/pages/roadmap/technology-tree.tsx]: Sure, have addressed in https://github.com/hashintel/hash/pull/2747/commits/2b4dc2588953fdcadf9a94e04f8be65c0d50efa1

File: apps/hashdotdev/src/pages/roadmap/technology-tree-data.tsx
Comment (nonparibus) [reply apps/hashdotdev/src/pages/roadmap/technology-tree-data.tsx:10]: It might be useful. Suggest we leave to hedge.

File: apps/hashdotdev/src/pages/roadmap/technology-tree-data.tsx
Comment (nonparibus) [reply apps/hashdotdev/src/pages/roadmap/technology-tree-data.tsx:470]: Let's do that tomorrow, or Wed, and aim to merge in subsequently.

File: apps/hashdotdev/src/pages/roadmap/technology-tree.tsx
Comment (nonparibus) [reply apps/hashdotdev/src/pages/roadmap/technology-tree.tsx:326]: Happy to sub out this behavior. It's not clear to me it's actually useful in its current form.

File: apps/hashdotdev/src/pages/roadmap.page.tsx
Comment (nonparibus) [reply apps/hashdotdev/src/pages/roadmap.page.tsx]: (In the above screenshot). At other screen sizes others suffer the same fate. image

**Comment by nonparibus** [comment]
Merge of this is also blocked by adding accurate use cases per node.

File: apps/hashdotdev/src/pages/roadmap/use-cases.tsx
Comment (benwerner01) [reply apps/hashdotdev/src/pages/roadmap/use-cases.tsx]: Ah great suggestion, have added in https://github.com/hashintel/hash/pull/2747/commits/8be978254c5d5035a65ffcfc16d7c0bc6eba4081

File: apps/hashdotdev/src/pages/roadmap/technology-tree.tsx
Comment (benwerner01) [reply apps/hashdotdev/src/pages/roadmap/technology-tree.tsx:326]: Happy to do this, but do we still want to tooltip to display on hover? Or on click as well?

File: apps/hashdotdev/src/pages/roadmap/technology-tree.tsx
Comment (CiaranMn) [reply apps/hashdotdev/src/pages/roadmap/technology-tree.tsx:362]: I would update the comment to say that we are aiming to avoid duplicate link elements, instead of saying it's avoiding duplicate react keys – the way to fix duplicate React keys is to give each element a distinct key. But it's the element themselves we don't want

Comment (CiaranMn) [reply apps/hashdotdev/src/pages/roadmap/technology-tree.tsx:362]: The error happens to be surfaced by a React warning, but even if there were no such thing as React keys we still wouldn't want multiple of the same link sitting on top of each other

File: apps/hashdotdev/src/pages/roadmap/technology-tree-data.tsx
Comment (CiaranMn) [reply apps/hashdotdev/src/pages/roadmap/technology-tree-data.tsx:10]: Let's update (or not) after we've agreed on the copy. It's better not to have files lying around called `tsx` that don't have any components in them because it makes it harder to understand the contents of the folder

File: apps/hashdotdev/src/pages/roadmap.page.tsx
Comment (benwerner01) [reply apps/hashdotdev/src/pages/roadmap.page.tsx]: Have centred the content in https://github.com/hashintel/hash/pull/2747/commits/72678f4c81125812f5e3204a91293d9fa9683846

image

File: apps/hashdotdev/src/pages/roadmap/technology-tree.tsx
Comment (CiaranMn) [reply apps/hashdotdev/src/pages/roadmap/technology-tree.tsx:326]: I would lean towards onclick, because if the tooltip appears on hover there is nothing to indicate that something would happen if you clicked on it. Or doing something else entirely to make the paths highlighted

Comment (CiaranMn) [reply apps/hashdotdev/src/pages/roadmap/technology-tree.tsx:326]: Alternatively we could have some other icon for highlight, or make the entire node clickable (appreciate we want to be able to link to 'read more' also, but there could be a smaller icon/link within the node for that)

File: apps/hashdotdev/src/pages/roadmap/technology-tree.tsx
Comment (CiaranMn) [reply apps/hashdotdev/src/pages/roadmap/technology-tree.tsx]: Looks good, thank you

File: apps/hashdotdev/src/pages/roadmap/technology-tree.tsx
Comment (benwerner01) [reply apps/hashdotdev/src/pages/roadmap/technology-tree.tsx]: I've made it on click in https://github.com/hashintel/hash/pull/2747/commits/2f3abe141d23882d12a8d535b242a4a29f385b7a, but think a separate button might be more appropriate. There's also a bug where if you then zoom, the tooltip becomes misaligned with the node which may not be very straightforward to address.

File: apps/hashdotdev/src/pages/roadmap/technology-tree.tsx
Comment (benwerner01) [inline apps/hashdotdev/src/pages/roadmap/technology-tree.tsx:219]: @nonparibus as you suggested in Figma I've made the technology tree wider than the paragraph elements (up to `1500px`). Let me know if you'd like me to adjust it to something else.

Comment (benwerner01) [inline apps/hashdotdev/src/pages/roadmap/technology-tree.tsx:219]: @nonparibus as you suggested in Figma I've made the technology tree wider than the paragraph elements (up to 1500px). Let me know if you'd like me to adjust it to something else.

image

Comment (nonparibus) [reply apps/hashdotdev/src/pages/roadmap/technology-tree.tsx:219]: On the latest deployment the right-hand filters look different for me than in the screenshot above. Any idea why? 
Capture 2023-07-31 at 19 36 27@2x

Comment (benwerner01) [reply apps/hashdotdev/src/pages/roadmap/technology-tree.tsx:219]: Because at that screen width the filter would overflow the page if we were to stick it on the right of the graph. If you were to make your screen wider it should appear on the right when there is enough space for it.

File: apps/hashdotdev/src/pages/roadmap/technology-tree.tsx
Comment (nonparibus) [reply apps/hashdotdev/src/pages/roadmap/technology-tree.tsx]: I'd rather not introduce a new button, as it'll add clutter to an already crowded interface.

I think we can afford for this functionality to be slightly hard to discover. Its utility in the first place remains unclear to me.

File: apps/hashdotdev/src/pages/roadmap.page.tsx
Comment (nonparibus) [reply apps/hashdotdev/src/pages/roadmap.page.tsx:202]: This isn't meant to link to `github.com/hashintel/hash` but the hash.dev/docs _getting started_ page.

It can link to the GitHub page as a stop-gap, though. Let's just update the link on the open branch that adds the `hash.dev/docs` section.

File: apps/hashdotdev/src/pages/roadmap/technology-tree.tsx
Comment (benwerner01) [reply apps/hashdotdev/src/pages/roadmap/technology-tree.tsx]: So to be clear you want me make it on hover again @nonparibus?

File: apps/hashdotdev/src/theme/components/utils/mui-css-baseline-theme-options.ts
Comment (CiaranMn) [reply apps/hashdotdev/src/theme/components/utils/mui-css-baseline-theme-options.ts:203]: Can't figure out why 🤷‍♂️

File: apps/hashdotdev/src/pages/roadmap.page.tsx
Comment (benwerner01) [reply apps/hashdotdev/src/pages/roadmap.page.tsx]: Added the link in https://github.com/hashintel/hash/pull/2747/commits/bd8d6b789cb9beb95db5645b5a412ff9f5e28d7b

File: apps/hashdotdev/src/pages/roadmap/technology-tree.tsx
Comment (nonparibus) [reply apps/hashdotdev/src/pages/roadmap/technology-tree.tsx]: What? No. Leave as onClick. I'm just saying don't add another icon that adds extra clutter.

File: apps/hashdotdev/src/pages/roadmap/technology-tree-data.tsx
Comment (CiaranMn) [inline apps/hashdotdev/src/pages/roadmap/technology-tree-data.tsx:693]:
```suggestion
    body: "Move to a more efficient data format for storing simulation data (e.g. Parquet)",
```

**Review by CiaranMn — COMMENTED** [review]
Adding a few descriptions, for @nonparibus to consider

**Review by nonparibus — DISMISSED** [review]

**Review by CiaranMn — DISMISSED** [review]

**Review by nonparibus — DISMISSED** [review]

**Review by nonparibus — APPROVED** [review]

**Pull request merged**
